cmd/gateway-b2.go

```diff
@@ -0,0 +1,660 @@
+	return PartInfo{}, traceError(NotImplemented{})
+}
+
+// PutObjectPart puts a part of object in bucket, uses B2's LargeFile upload API.
+func (l *b2Objects) PutObjectPart(bucket string, object string, uploadID string, partID int, data *HashReader) (pi PartInfo, err error) {
+	bkt, err := l.Bucket(bucket)
+	if err != nil {
+		return pi, err
+	}
+
+	fc, err := bkt.File(uploadID, object).CompileParts(0, nil).GetUploadPartURL(l.ctx)
+	if err != nil {
+		return pi, b2ToObjectError(traceError(err), bucket, object, uploadID)
+	}
+
+	hr := newB2Reader(data, data.Size())
+	sha1, err := fc.UploadPart(l.ctx, hr, sha1AtEOF, int(hr.Size()), partID)
+	if err != nil {
+		return pi, b2ToObjectError(traceError(err), bucket, object, uploadID)
+	}
+
+	return PartInfo{
+		PartNumber:   partID,
+		LastModified: UTCNow(),
+		ETag:         sha1,
+		Size:         data.Size(),
+	}, nil
+}
+
+// ListObjectParts returns all object parts for specified object in specified bucket, uses B2's LargeFile upload API.
+func (l *b2Objects) ListObjectParts(bucket string, object string, uploadID string, partNumberMarker int, maxParts int) (lpi ListPartsInfo, err error) {
+	bkt, err := l.Bucket(bucket)
+	if err != nil {
+		return lpi, err
+	}
+	lpi = ListPartsInfo{
+		Bucket:           bucket,
+		Object:           object,
+		UploadID:         uploadID,
+		MaxParts:         maxParts,
+		PartNumberMarker: partNumberMarker,
+	}
+	// startPartNumber must be in the range 1 - 10000 for B2.
+	partNumberMarker++
+	partsList, next, err := bkt.File(uploadID, object).ListParts(l.ctx, partNumberMarker, maxParts)
+	if err != nil {
+		return lpi, b2ToObjectError(traceError(err), bucket, object, uploadID)
+	}
+	if next != 0 {
+		lpi.IsTruncated = true
+		lpi.NextPartNumberMarker = next
+	}
+	for _, part := range partsList {
+		lpi.Parts = append(lpi.Parts, PartInfo{
+			PartNumber: part.Number,
+			ETag:       part.SHA1,
+			Size:       part.Size,
+		})
+	}
+	return lpi, nil
+}
+
+// AbortMultipartUpload aborts a ongoing multipart upload, uses B2's LargeFile upload API.
+func (l *b2Objects) AbortMultipartUpload(bucket string, object string, uploadID string) error {
+	bkt, err := l.Bucket(bucket)
+	if err != nil {
+		return err
+	}
+	err = bkt.File(uploadID, object).CompileParts(0, nil).CancelLargeFile(l.ctx)
+	return b2ToObjectError(traceError(err), bucket, object, uploadID)
+}
+
+// CompleteMultipartUpload completes ongoing multipart upload and finalizes object, uses B2's LargeFile upload API.
+func (l *b2Objects) CompleteMultipartUpload(bucket string, object string, uploadID string, uploadedParts []completePart) (oi ObjectInfo, err error) {
+	bkt, err := l.Bucket(bucket)
+	if err != nil {
+		return oi, err
+	}
+	hashes := make(map[int]string)
+	for _, uploadedPart := range uploadedParts {
+		hashes[uploadedPart.PartNumber] = uploadedPart.ETag
+	}
+	_, err = bkt.File(uploadID, object).CompileParts(0, hashes).FinishLargeFile(l.ctx)
+	if err != nil {
+		return oi, err
```

Comment:
shouldn't we call `b2ToObjectError(traceError(err), ...)`?

Reply:
done